test(dang-workflows-remediation): create initial tests

diogoteles08 opened the pull request:
Tests are covering the overall run of the dangerous workflow script injection fix -- inputs as the whole workflow and outputs as the fixed workflow.

diogoteles08 committed:
test(dang-workflows-remediation): create initial tests

Tests are covering the overall run of the dangerous workflow script
injection fix -- inputs as the whole workflow and outputs as the fixed
workflow.

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
fix: fix linter issues and add LICENSE header on new files

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

pnacht reviewed (changes requested):
Move inputs and outputs to dedicated files, and only have the filepaths here. This is already done for the DangerousWorkflow check, for example:

https://github.com/joycebrum/scorecard/blob/main/checks/raw/dangerous_workflow_test.go#L110

I believe this:
- makes it easier to visually inspect the "diffs" between the files (using VSCode or `git diff`, for example)
- makes it easier to visually "parse" the individual files, since we can use VSCode's syntax highlighting
- makes the test logic more legible, since `impl_test` goes from "90% examples" to "90% logic".

You can also store only the input filename, and the expected output filename is simply "fixed_{input}" or something.

diogoteles08 commented:
> Move inputs and outputs to dedicated files, and only have the filepaths here. This is already done for the DangerousWorkflow check, for example:
> 
> https://github.com/joycebrum/scorecard/blob/main/checks/raw/dangerous_workflow_test.go#L110
> 
> I believe this:
> 
> * makes it easier to visually inspect the "diffs" between the files (using VSCode or `git diff`, for example)
> * makes it easier to visually "parse" the individual files, since we can use VSCode's syntax highlighting
> * makes the test logic more legible, since `impl_test` goes from "90% examples" to "90% logic".
> 
> You can also store only the input filename, and the expected output filename is simply "fixed_{input}" or something.

I have mixed feelings about this, because I agree with the point of splitting the test logic from the examples themselves, and I like to have the examples outside the actual code, but the current version is way easier to me to read. My arguments:
1. The examples on this file are small and very straightforward -- I think it'd be valuable to have extensive and real workflows as a different test as well, and those definitely should be in different files.
2. The logic of this test is barely dumb: I'm only making the diff and checking if it's empty. So I could argue that a spending a whole file for this logic could be unnecessary.
3. It's **so easier** to me to read the test that is self-contained on a single file. E.g., when I followed [your link](https://github.com/joycebrum/scorecard/blob/main/checks/raw/dangerous_workflow_test.go#L110) to understand the test you mentioned, I had to spend some time reading the test file to find the actual path for the files (bc the path on the test was a relative path), and had to be changing from file to file to actually understand the test, and that really annoys me (too many tabs! aaaa).

That said, one purpose that I have is to just update this file to 
1. make the examples even more straightforward -- remove the workflow names, triggers, etc; keep only the actual job that we're evaluating. 
2. Create a second test function (or test file, if you prefer), to run tests over whole real workflows, extracted from real repositories -- and them put the workflows as separated files and refer to them by filepath.

WDYT?

joycebrum commented:
Humm this is a good point. I'd say for us to follow the "dedicated file" option to be in accordance to how scorecard is doing it now. 

Besides, I'd also like to see the tests running on very complex cases -- with lots of different identations and comments, multiple script injections on the same file, etc. We could even use the existing ones they use for testing script injection identification and just create the "expected" for it.

pnacht commented:
+1 to what Joyce said.

diogoteles08 commented:
Ok, I'm working on it. But a question that came up:
For the workflow examples that I'm making up, I'm adding the header of the Scorecard LICENSE (as it was added on the example you sent me), but does it make sense to also add it on the workflow that I'm just copying from a real scenario? For example, I'm including as a test example a workflow that SACI has fixed for Angular. 

Additional tricky question: is it OK to mention that the workflow was extracted from Angular?

pnacht commented:
I don't think we need to add licenses to the test files. There are already some test cases which don't have licenses:

https://github.com/joycebrum/scorecard/blob/main/checks/raw/testdata/.github/workflows/github-pysa-workflow.yaml


As for mentioning the workflow came from Angular... I think it'd be fine, but I wouldn't bother: it doesn't add any _relevant_ information about the test; we care about the behavior expressed by the test case, not its origins.

diogoteles08 committed:
move test data to external files, add new tests

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
Merge branch 'main' into test/initial-tests-for-dw-fix

diogoteles08 commented:
Hey @joycebrum and @pnacht, I've made the changes you requested, added some additional tests, and merged the code on main into my branch. PTAL =)

pnacht commented on probes/hasDangerousWorkflowScriptInjection/patch/testdata/badIndentationMultipleInjections.yaml:
```diff
@@ -0,0 +1,36 @@
+# limitations under the License.
+  on: [pull_request]
+
+jobs:
+      build:
+    name: Build and test
```
This isn't valid GitHub Action syntax. Remember that YAML requires consistent indentation. If `build` is a job name (the thing expected under `jobs`), then what is `name`? It's in a separate indentation level but not the same indentation as `jobs`, so it's not in the global scope either.

I'm 100% in favor of bad indentation to make sure the remediation keeps the existing style, but I don't think it needs to be tested on an invalid workflow (unless we want to check that it "fails gracefully").

diogoteles08 replied:
Hey Pedro! No, I just didn't know that this would fail. Given that workflows with bad indentation work, I was assuming it didn't care much about indentation.

So, how can I place the indentation in order to make it a bad indentation, but still run on GitHub? Make Indentation with arbitrary number of Tabs instead of always 1, but still keeping them consistent between themselves?

diogoteles08 replied:
I'm taking a look at https://learnxinyminutes.com/docs/yaml/ and I think I'm already able to follow on this. I'll let you know if I still need help

pnacht commented on probes/hasDangerousWorkflowScriptInjection/patch/testdata/twoInjectionsDifferentJobs.yaml:
```diff
@@ -0,0 +1,28 @@
+
+jobs:
+	fascinating-job:
+		steps:
```
I notice your synthetic examples all use tabs for indentation. That's a good test to make sure the remediation keeps that syntax (it currently doesn't, in this case!), but note that the vast majority of real workflows use 2-space indentation instead.

diogoteles08 replied:
Hmm nice point. I'll change the indentation of my tests to use 2-space, to follow this pattern -- but on the test about Indentation, I'll mix the 2 types of indentation and also test this out hehe

diogoteles08 replied:
FWI, by [this documentation](https://learnxinyminutes.com/docs/yaml/)  and this [YAML validator](https://www.yamllint.com), tab-indentation is actually not even allowed/recognized on official YAML

pnacht replied:
TIL! That's neat! Also makes a lot of sense (imagine a doc that mixes spaces and tabs... the "structure" of the doc will appear different to people who use tabs = 2 or 4 spaces).

pnacht commented on probes/hasDangerousWorkflowScriptInjection/patch/testdata/safeExample.yaml:
```diff
@@ -0,0 +1,24 @@
+    runs-on: ubuntu-latest
+    steps:
+      - run: |
+          echo "Comment on issue #${{ github.event.issue.number }}"
```
Excellent testcase! Though I'd rename it to make it clear it's testing comments

diogoteles08 replied:
Nice point!

pnacht reviewed (changes requested):
Additional tests required:
- Workflow already has an `env`
  - and a variable must be appended to the env; OR
  - the env is "irrelevant", variables must be added at a different scope
- Workflow ending with a newline (all the current examples don't)
- Results when the script is run to add all environment variables at the workflow level, not at the smallest scope.

diogoteles08 committed:
test(dw-script-injections): add test for newline before EOF

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(dw-scipt-injection): improve naming of test

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injections): change tab indentation to 2-space

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injections): fix and enhance indentation test

Makes the test file a valid YAML file, but also enhances it to also
cover:
1. the case of having a newline at EOF
2. test if the patch will keep a tab character present in a comment

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injection): add emmpty lines between all blocks in root level

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injections): move all env vars to root level

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injection): specific test for workflows without blanklines

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injection): test case of env vars already defined

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
fix(fix-script-injections): fix linter issues

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

pnacht commented on probes/hasDangerousWorkflowScriptInjection/patch/testdata/twoInjectionsSameStep.yaml:
```diff
@@ -0,0 +1,25 @@
+      - name: where things are done
+        run: |
+          echo "${{ github.event.issue_comment.comment }}"
+          mkdir "${{ github.event.fork.forkee.name }}"
```
`fork.forkee.name` isn't a dangerous input. It is user-controlled, but the rules for repo names make it impossible to craft a malicious name (no spaces, no slashes, etc).

pnacht commented on probes/hasDangerousWorkflowScriptInjection/patch/testdata/twoInjectionsSameJob_fixed.yaml:
```diff
@@ -0,0 +1,30 @@
+    types: [created]
+
+env:
+  DISCUSSION_BODY: ${{ github.event.discussion.body }}
+  DISCUSSION_TITLE: ${{ github.event.discussion.title }}
```
Good one! I don't think Scorecard actually detects this... please suggest it in an issue there!

diogoteles08 replied:
https://github.com/ossf/scorecard/issues/3915

pnacht commented on probes/hasDangerousWorkflowScriptInjection/patch/testdata/userInputAssignedToVariable_fixed.yaml:
```diff
@@ -0,0 +1,34 @@
+on: [pull_request]
+
+env:
+  COMMITS_MESSAGE: ${{ github.event.commits[0].message }}
```
The current implementation calls this "COMMIT_MESSAGE". That can lead to a conflict with `head_commit.message`, but I think that's fine for v1...

It's probably a rare conflict, and "COMMITS_MESSAGE" sounds weird, since it's the message of a single commit.

pnacht commented on probes/hasDangerousWorkflowScriptInjection/patch/testdata/realExample3_fixed.yaml:
```diff
@@ -0,0 +1,47 @@
+      - '**/CHANGELOG.md'
+
+env:
+  PULL_REQUEST_BODY: ${{ github.event.pull_request.body }} 
```
For all the `pull_request.*` variables, I've shortened the envvar names to "PR_*"

pnacht commented on probes/hasDangerousWorkflowScriptInjection/patch/testdata/envVarsAlreadyExist_fixed.yaml:
```diff
@@ -0,0 +1,84 @@
+permissions: read-all
+
+jobs:
+  using workflow-level env vars:
```
Job names can't have spaces, only alphanumeric, "-" and "_".

pnacht commented on probes/hasDangerousWorkflowScriptInjection/patch/impl_test.go:
```diff
@@ -13,3 +13,107 @@
+			expectedFilepath: "noLineBreaksBetweenBlocks_fixed.yaml",
+		},
+		{
+			name:             "Ignore if user input regex is just part of a comment",
+			inputFilepath:    "safeExample.yaml",
+			expectedFilepath: "safeExample.yaml",
```
Update to new filename.

pnacht commented on probes/hasDangerousWorkflowScriptInjection/patch/impl_test.go:
```diff
@@ -13,3 +13,107 @@
+			inputFilepath:    "badIndentationMultipleInjections.yaml",
+			expectedFilepath: "badIndentationMultipleInjections_fixed.yaml",
+		},
+		{
+			// Currently we're not keeping this pattern, as we always add a blankline after the env block
+			name:             "File with no blank lines between blocks",
+			inputFilepath:    "noLineBreaksBetweenBlocks.yaml",
+			expectedFilepath: "noLineBreaksBetweenBlocks_fixed.yaml",
+		},
```
In such cases, where we know the current behavior isn't satisfactory, we can have two tests: one with what it SHOULD be (commented out), and one with how it is:

```suggestion
		{
			// Currently we're not keeping this pattern, as we always add a blankline after the env block
			name:             "File with no blank lines between blocks",
			inputFilepath:    "noLineBreaksBetweenBlocks.yaml",
			expectedFilepath: "noLineBreaksBetweenBlocks_fixed.yaml",
		},
		// {
			// Test that currently fails due to lack of style awareness.
			// Once proper behavior is implemented, enable this test
			// and delete the one above.
			// name:             "File with no blank lines between blocks",
			// inputFilepath:    "noLineBreaksBetweenBlocks.yaml",
			// expectedFilepath: "noLineBreaksBetweenBlocks_fixed.yaml",
		// },
```

diogoteles08 replied:
I like the idea of having the commented one and I'm doing that, but is the "current" one really necessary? Theoretically I think it doesn't test anything that was not covered on other cases. And we would already have the commented one to raise awareness on how it should be -- I can also mention on the comment that we currently always add a blank line

pnacht replied:
Well, having the "current" test enabled ensures there are no regressions. But I agree, if the test doesn't test anything that's not covered by other tests already, it can be dropped.

pnacht commented on probes/hasDangerousWorkflowScriptInjection/impl.go:
```diff
@@ -63,8 +63,8 @@ func Run(raw *checker.RawResults) ([]finding.Finding, string, error) {
 			})
-			patch := patch.GeneratePatch(e.File)
-			f.WithPatch(&patch)
+			finding_patch := patch.GeneratePatch(e.File)
+			f.WithPatch(&finding_patch)
```
Is this relevant for this PR?

diogoteles08 replied:
Ah it was a linter fix. This was shadowing the package name. Not strictly related to PR, but I'd keep it here just to keep everything linted afterwareds

joycebrum replied:
Just a thing I've noticed now: I think the pattern here should be CamelCase instead of snaked case.

```suggestion
			findingPatch := patch.GeneratePatch(e.File)
			f.WithPatch(&findingPatch)
```

diogoteles08 replied:
Oh I think you're right! It's weird that the linter didn't notice that, though 😕

pnacht reviewed (changes requested):
.

diogoteles08 committed:
test(fix-script-injections): split tests with existent env vars

Now we have passing tests, and commented tests that require future
modifications on our implementation.

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injections): also test inputs on discussion event

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injections): remove fork.forkee example because it's safe

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injections): change test with no blank lines to corresponde with the ideal

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injections): fixed testdata was actually not fixed

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injections): change env var names to be compatible to our script

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injections): fix typo on file name

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

gabibguti commented:
One comment about "testing complex indentation cases". Indentation is critical for YAML structure, it defines, for example, if a variable is inside one key or another. Given that, I don't think it makes sense to test "badIndentationMultipleInjections.yaml". Testing multiple injections is a good case, but changing the indentation no. At least, I don't agree.

gabibguti commented:
Also, we should fix the Copyright notices date from 2021 to 2024.

diogoteles08 commented:
> One comment about "testing complex indentation cases". Indentation is critical for YAML structure, it defines, for example, if a variable is inside one key or another. Given that, I don't think it makes sense to test "badIndentationMultipleInjections.yaml". Testing multiple injections is a good case, but changing the indentation no. At least, I don't agree.

Take a look at [this](https://github.com/joycebrum/scorecard/pull/2#discussion_r1511822330) and [this](https://github.com/joycebrum/scorecard/pull/2#discussion_r1511825258) threads. I acknowledge that indentation is critical for YAML, and I think this makes this test even more important. The workflow on this test case is perfectly valid, and we've already found real workflows with messed up indentation as this one, so we must evaluate how our tool handles these cases.

The way we chose to handle this case is to leave the indentation as it is -- this avoids unnecessary diffs and doesn't mess with maintainer's authority on code style (idk, they might have a reason to let it as it is). 

Sure, we're open to revisit this later and say that we should fix the indentation anyways, who knows. But this tests would still be important to test out this scenario.

diogoteles08 committed:
test(fix-script-injections): split badIndentation test into clearer ones

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injections): update copyright notices to 2024

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

pnacht commented on probes/hasDangerousWorkflowScriptInjection/patch/testdata/fourSpacesIndendationExistentEnvVar.yaml:
```diff
@@ -0,0 +1,28 @@
+# Copyright 2021 OpenSSF Scorecard Authors
```
These examples are inverted, this is the fixed one.

diogoteles08 committed:
test(fix-script-injections): fix typo on file names

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injections): fix linter issues

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

joycebrum reviewed (approved):
LGTM 
We can decide to remove the commented tests later, let's keep it there by now.

diogoteles08 committed:
convert snake_case variable to camelCase

Co-authored-by: Joyce <joycebrum@google.com>
Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

gabibguti commented:
What is the goal of "ignorePatternInsideComments.yaml"?

gabibguti commented on probes/hasDangerousWorkflowScriptInjection/patch/testdata/noLineBreaksBetweenBlocks.yaml:
```diff
@@ -0,0 +1,24 @@
+jobs:
+  simple-job:
+    steps:
+      - name: As simple as it cam be
```
Typo
```suggestion
      - name: As simple as it can be
```

gabibguti commented on probes/hasDangerousWorkflowScriptInjection/patch/testdata/noLineBreaksBetweenBlocks_fixed.yaml:
```diff
@@ -0,0 +1,26 @@
+jobs:
+  simple-job:
+    steps:
+      - name: As simple as it cam be
```
Typo
```suggestion
      - name: As simple as it can be
```

joycebrum commented on probes/hasDangerousWorkflowScriptInjection/patch/impl_test.go:
```diff
@@ -13,3 +13,136 @@
+			}
+
+			output := GeneratePatch(inputFile)
+			if diff := cmp.Diff(string(expectedContent[:]), output); diff != "" {
+				t.Errorf("mismatch (-want +got):\n%s", diff)
+			}
```
Can we comment this part to avoid tests to break? Once it gets to main, all the future PRs won't be clear if they are breaking something or not.

```suggestion
			// if diff := cmp.Diff(string(expectedContent[:]), output); diff != "" {
				// t.Errorf("mismatch (-want +got):\n%s", diff)
			// }
```

joycebrum reviewed (changes requested):
Just noticed that the tests not passing could be a problem in the future PRs.

diogoteles08 committed:
Fix typo on text

Co-authored-by: Gabriela Gutierrez <gabigutierrez@google.com>
Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
fix typo on text of fixed file

Co-authored-by: Gabriela Gutierrez <gabigutierrez@google.com>
Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 committed:
test(fix-script-injection): comment line that warns test failure

We will uncomment it when the code is actually ready to be tested.

Signed-off-by: Diogo Teles Sant'Anna <diogoteles@google.com>

diogoteles08 commented:
> What is the goal of "ignorePatternInsideComments.yaml"?

It just evaluates if the script isn't changing code that is part of a comment, as it wouldn't be unsafe.
